File: Algorithms/0036_Valid_Sudoku/Python/Valid_Sudoku_Solution_1.py

# Space: O(n)
# Time: O(n)

class Solution:
    def isValidSudoku(self, board) -> bool:

        # check row
        if not self.check_line(board, 0): return False

        # check column
        if not self.check_line(board, 1): return False

        # check box
        start_point = [[0, 0], [0, 3], [0, 6], [3, 0], [3, 3], [3, 6], [6, 0], [6, 3], [6, 6]]
        for y, x in start_point:
            if not self.check_box(board, x, y): return False

        return True

    def check_line(self, board, flag):
        if flag == 0:  # check row
            for y in range(9):
                cache = set()
                for x in range(9):
                    if board[y][x] == '.':
                        continue
                    elif board[y][x] not in cache:
                        cache.add(board[y][x])
                    else:
                        return False

        else:  # check column
            for x in range(9):
                cache = set()
                for y in range(9):
                    if board[y][x] == '.':
                        continue
                    elif board[y][x] not in cache:
                        cache.add(board[y][x])
                    else:
                        return False
        return True

    def check_box(self, board, x_start, y_start):
        cache = set()
        for y in range(y_start, y_start + 3):
            for x in range(x_start, x_start + 3):
                if board[y][x] == '.':
                    continue
                elif board[y][x] not in cache:
                    cache.add(board[y][x])
                else:
                    return False
        return True

File: Algorithms/0036_Valid_Sudoku/Java/ValidSudokuSolution1.java

// Space: O(n)
// Time: O(n)

class Solution {
    public boolean isValidSudoku(char[][] board) {
        int[][] startPoint = new int[][]{{0, 0}, {0, 3}, {0, 6}, {3, 0}, {3, 3}, {3, 6}, {6, 0}, {6, 3}, {6, 6}};
        if (!checkLine(board, 1)) return false;
        if (!checkLine(board, 0)) return false;
        for (int[] temp : startPoint) {
            if (!checkBox(board, temp[0], temp[1])) return false;
        }
        return true;
    }

    private boolean checkLine(char[][] board, int flag) {
        if (flag == 1) {  // check row
            for (int y = 0; y < 9; y++) {
                Set<Character> cache = new HashSet<>();
                for (int x = 0; x < 9; x++) {
                    if (board[y][x] == '.') continue;
                    if (!cache.contains(board[y][x])) {
                        cache.add(board[y][x]);
                    } else return false;
                }
            }
        } else { // check column
            for (int x = 0; x < 9; x++) {
                Set<Character> cache = new HashSet<>();
                for (int y = 0; y < 9; y++) {
                    if (board[y][x] == '.') continue;
                    if (!cache.contains(board[y][x])) {
                        cache.add(board[y][x]);
                    } else return false;
                }
            }
        }
        return true;
    }

    private boolean checkBox(char[][] board, int yStrat, int xStart) {
        Set<Character> cache = new HashSet<>();
        for (int y = yStrat; y < yStrat + 3; y++) {
            for (int x = xStart; x < xStart + 3; x++) {
                if (board[y][x] == '.') continue;
                if (!cache.contains(board[y][x])) {
                    cache.add(board[y][x]);
                } else return false;
            }
        }
        return true;
    }
}





